src/awkward/_connect/numba/layoutbuilder.py

```diff
@@ -0,0 +1,2547 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+
+import math
+
+import numba
+import numba.core.typing.npydecl
+import numpy as np
+
+import awkward as ak
+from awkward._connect.numba.growablebuffer import GrowableBuffer, GrowableBufferType
+from awkward._nplikes.numpylike import ArrayLike
+from awkward._typing import final
+
+
+class LayoutBuilder:
+    def __init__(self, builder, *, parameters=None):
+        self._builder = builder
+        self._parameters = parameters
+
+    @property
+    def builder(self):
+        return self._builder
+
+    def clear(self):
+        return self._builder.clear()
+
+    def is_valid(self, error: str):
+        return self._builder.is_valid(error)
+
+    def _type(self, typestrs):
+        raise NotImplementedError
+
+
+def to_numbatype(builder):
+    if isinstance(builder, Numpy):
+        return Numpy.numbatype(builder)
+    elif isinstance(builder, Empty):
+        return Empty.numbatype(builder)
+    elif isinstance(builder, ListOffset):
+        return ListOffset.numbatype(builder)
+    elif isinstance(builder, Regular):
+        return Regular.numbatype(builder)
+    elif isinstance(builder, IndexedOption):
+        return IndexedOption.numbatype(builder)
+    elif isinstance(builder, ByteMasked):
+        return ByteMasked.numbatype(builder)
+    elif isinstance(builder, BitMasked):
+        return BitMasked.numbatype(builder)
+    elif isinstance(builder, Unmasked):
+        return Unmasked.numbatype(builder)
+    elif isinstance(builder, Record):
+        return Record.numbatype(builder)
+    elif isinstance(builder, Tuple):
+        return Tuple.numbatype(builder)
+    elif isinstance(builder, Union):
+        return Union.numbatype(builder)
+    elif isinstance(builder, tuple):
+        return numba.types.Tuple([to_numbatype(it) for it in builder])
+    else:
+        return builder
+
+
+class LayoutBuilderType(numba.types.Type):
+    def __init__(self, parameters):
+        super().__init__(name="ak.numba.lb.LayoutBuilder()")
+
```

Comment:
This is another abstract class; don't define an `__init__`.

```suggestion
    def _init(self, parameters):
        self._parameters = parameters
```